File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) 
    {
        int length = 0;
        ListNode* curr = head;
        while (curr) {
            length++;
            curr = curr->next;
        }

        if (k > length || k <= 0) {
            return head;
        }

        if (k == length - k + 1) {
            return head;
        }

        curr = head;
        for (int i = 1; i < k; i++) {
            curr = curr->next;
        }
        ListNode* first_node = curr;

        curr = head;
        for (int i = 1; i < length - k + 1; i++) {
            curr = curr->next;
        }
        ListNode* second_node = curr;

        swap(first_node->val, second_node->val);

        return head;
    }
};

File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next

        # Handle edge cases
        if k > length:
            return head

        if k == length - k + 1:
            return head

        curr = head
        for i in range(1, k):
            curr = curr.next
        first_node = curr

        curr = head
        for i in range(1, length - k + 1):
            curr = curr.next
        second_node = curr

        # Swap the values of the two nodes
        first_node.val, second_node.val = second_node.val, first_node.val
        return head